Make prefer-stateless-function PureComponent aware

tiemevanveen opened the pull request:
prefer-stateless-function should not return a warning when the user uses `PureComponent` to opt in for free optimisations with `shouldComponentUpdate` and is using props or context. 

Current false positive: 

```{js}
class Foo extends React.PureComponent {
  render() {
    return <div>{this.props.foo}</div>;
  }
}
```

I first only added the same `isPureComponent` check as [require-optimization](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/require-optimization.js#L92) does. But realized that components that are extending form PureComponent but don't use `this.props` or `this.context` should be written as a pure function as well. So I had to change the existing checks a bit to check that too. 

Yeah I know, I could have first reacted to the discussion at #777, but I thought it would be cool to find out how all the 'rules' in eslint work by trying to fix this issue. Hope you agree about the false positive. I think this case is different from #777 because that case it not using props or context at all.

Please correct any formatting fails, first contribution here.

lencioni commented on lib/rules/prefer-stateless-function.js:
```diff
@@ -65,6 +69,19 @@ module.exports = {
+      }
+
+      return false;
+    };
```
I wonder if this function should go in lib/util/Components.js

ljharb replied:
Agreed

lencioni commented:
This seems like a good change to me. I think it could be DRYed up a little with some of the code borrowed from other rules.

The only question I have is--should this be the behavior for everyone, or should it be an option? I'm inclined to think it can be the behavior for everyone for the sake of simplicity, but I'd like to hear opinions on that.

tiemevanveen commented:
Thanks for the quick response. 

Yeah I changed as little as possible. Just let me know if you want me to move some of it to the utils. That makes sense.

ljharb commented on docs/rules/prefer-stateless-function.md:
```diff
@@ -8,10 +8,11 @@ This rule will check your class based React components for
 * `render` method that return anything but JSX: `undefined`, `null`, etc. (only in React <15.0.0, see [shared settings](https://github.com/yannickcr/eslint-plugin-react/blob/master/README.md#configuration) for React version configuration)
 
-If none of these 4 elements are found, the rule will warn you to write this component as a pure function.
+If none of these 5 elements are found, the rule will warn you to write this component as a pure function.
```
Let's just make this say "these elements" to avoid having to update the count in the future.

ljharb commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -43,3 +44,14 @@ class Foo extends React.Component {
 }
 ```
+
+The following pattern is not considered warnings:
````
a warning

tiemevanveen replied:
haha yeah that's what I wrote, but the rest of docs we're like this so I didn't change it. Will fix this.

ljharb commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -43,3 +44,14 @@ class Foo extends React.Component {
+The following pattern is not considered warnings:
+
+```js
+class Foo extends React.PureComponent {
````
I still think this pattern should be considered a warning - being a PureComponent does not mean this is better as a stateful component.

tiemevanveen replied:
Being a PureComponent does not make it a stateful component right? It just makes it benefit from `shouldComponentUpdate`

ljharb replied:
Being a PureComponent does not make it stateless either - the two things are orthogonal. I'm saying `prefer-stateless-function` is far more important than `prefer-optimization`.

ljharb commented:
I'm -1 on this, per #777.

tiemevanveen commented:
on #777 
> You should be preferring a stateless function there, since there's no state. "PureComponent" usage doesn't affect this rule by design.

But if the user chose to use PureComponent instead of Component and uses props or context, doesn't that indicate that the user want's the component to only update if the props or context changes? And that's not possible with a stateless component.

And if this is not desired then there should also be a warning when the user implemented `shouldComponentUpdate` or used the `PureRenderMixin` right? And both are not triggering a warning now.

ljharb commented:
If the user wants to convey their intentions to ignore this rule that prefers components without state be SFCs, they can do that with an eslint override.

While I agree that implementing `shouldComponentUpdate` means that the component can't be blindly made into an SFC, I think that if a `React.createClass` implements `PureRenderMixin`, but in all other respects should be an SFC - that it should be a warning.

lencioni commented:
Yeah I was originally thinking that using PureComponent was a strong
indication that the developer prefers this over an SFC, but taking a step
back I don't think that's accurate. I agree that this should be handled on
a case by case basis with an eslint override and not by weakening this rule.

On Sun, Aug 21, 2016, 9:59 AM Jordan Harband <notifications@github.com>
wrote:

> If the user wants to convey their intentions to ignore this rule that
> prefers components without state be SFCs, they can do that with an eslint
> override.
>
> While I agree that implementing shouldComponentUpdate means that the
> component can't be blindly made into an SFC, I think that if a
> React.createClass implements PureRenderMixin, but in all other respects
> should be an SFC - that it should be a warning.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/yannickcr/eslint-plugin-react/pull/781#issuecomment-241268825>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAL7znaeECA0K2h02pjVFCexPL1cgmKBks5qiIP2gaJpZM4JpR5e>
> .
>

tiemevanveen commented:
Okay, fair enough.

But currently it's not possible to globally override this rule in `.eslintrc` so that it doesn't trigger a warning when PureComponent is used right?

ljharb commented:
@tiemevanveen that's true - making an option to `prefer-stateless-function` like "ignorePureComponents" might be worth doing, however?

tiemevanveen commented:
Added the option. Also moved `isPureComponent(node)` to the utils in `Components.js`.

tiemevanveen commented on lib/util/Components.js:
```diff
@@ -200,9 +200,22 @@ function componentRule(rule, context) {
      * Check if the node is returning JSX
      *
-     * @param {ASTNode} ASTnode The AST node being checked
+     * @param {ASTNode} node The AST node being checked
```
all the other methods say 'node'..

lencioni replied:
This token should match the argument name, so you should update line 222 at the same time.

tiemevanveen replied:
Check. And that's not possible because node is reassigned afterwards. Changed back tot original value. Thanks.

ljharb commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -43,3 +44,14 @@ class Foo extends React.Component {
 }
 ```
+
+The following pattern is not considered a warning:
````
I think this documentation needs to be updated to describe the option?

ljharb commented on lib/rules/prefer-stateless-function.js:
```diff
@@ -244,6 +276,12 @@ module.exports = {
 
     return {
+      ClassDeclaration: function (node) {
+        if (utils.isPureComponent(node) && ignorePureComponents === true) {
```
I don't think we need the `=== true` :-) also we should probably check the boolean first to avoid the possibly unnecessary isPureComponent call.

tiemevanveen commented:
You're right, thanks. 

Fixed and added to last commit. 

Also removed the `=== true` form 'no-multi-comp' where I stole the example from in the first place ;)

yannickcr commented on docs/rules/prefer-stateless-function.md:
```diff
@@ -43,3 +44,36 @@ class Foo extends React.Component {
+
+### `ignorePureComponent`
+
+When `true` the rule will ignore Components extending from `React.PureComponent` that use `this.props` or `this.context`.
```
Can you indicate the default value here please?

tiemevanveen replied:
Done (added to commit).

tiemevanveen replied:
forgot to push.. done now

lencioni commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -43,3 +44,36 @@ class Foo extends React.Component {
+
+The following patterns is considered okay and does not cause warnings:
+
+```js
````
nit: please make these `jsx` instead of `js`, for best syntax highlighting

tiemevanveen replied:
done

yannickcr commented:
Merged, thanks!

tiemevanveen commented:
Thanks!

swarthy commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -45,13 +45,38 @@ class Foo extends React.Component {
 
 ```js
+...
+"prefer-stateless-function": [<enabled>, { "ignorePureComponent": <ignorePureComponent> }]
````
ignorePureComponents not ignorePureComponent

ljharb replied:
Thanks, fixed in 9f764599a537728e7ac9fa332d0ac6e3a6deb788

swarthy commented on docs/rules/prefer-stateless-function.md:
```diff
@@ -45,13 +45,38 @@ class Foo extends React.Component {
+* `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
+* `ignorePureComponent`: optional boolean set to `true` to ignore components extending from `React.PureComponent` (default to `false`).
+
+### `ignorePureComponent`
```
ignorePureComponents

ljharb replied:
Thanks, fixed in 9f764599a537728e7ac9fa332d0ac6e3a6deb788

swarthy commented on docs/rules/prefer-stateless-function.md:
````diff
@@ -45,13 +45,38 @@ class Foo extends React.Component {
+```
+
+* `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
+* `ignorePureComponent`: optional boolean set to `true` to ignore components extending from `React.PureComponent` (default to `false`).
````
ignorePureComponents

ljharb replied:
Thanks, fixed in 9f764599a537728e7ac9fa332d0ac6e3a6deb788
